sql/analyzer/resolve_subqueries.go

```diff
@@ -110,64 +105,83 @@ func flattenTableAliases(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope
+		// Specifically, if the subquery isn't resolved or a column can't be found in the scope node, wait until a later pass.
+		if !finalize && (ErrValidationResolved.Is(err) || sql.ErrTableColumnNotFound.Is(err) || sql.ErrColumnNotFound.Is(err)) {
+			// keep the work we have and defer remainder of analysis of this subquery until a later pass
+			return sq.WithQuery(analyzed), transform.NewTree, nil
 		}
+		return nil, transform.SameTree, err
+	}
 
-		//todo(max): Infinite cycles with subqueries, unions, ctes, catalog.
-		// we squashed most negative errors, where a rule fails to report a plan change
-		// to the expense of positive errors, where a rule reports a change when the plan
-		// is the same before/after.
-		// .Resolved() might be useful for fixing these bugs.
-		return s.WithQuery(StripPassthroughNodes(analyzed)), transform.NewTree, nil
-	})
+	//todo(max): Infinite cycles with subqueries, unions, ctes, catalog.
+	// we squashed most negative errors, where a rule fails to report a plan change
+	// to the expense of positive errors, where a rule reports a change when the plan
+	// is the same before/after.
+	// .Resolved() might be useful for fixing these bugs.
+	return sq.WithQuery(StripPassthroughNodes(analyzed)), transform.NewTree, nil
 }
 
-func finalizeSubqueryExpressions(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, sel RuleSelector) (sql.Node, transform.TreeIdentity, error) {
-	return transform.NodeExprsWithNode(n, func(n sql.Node, e sql.Expression) (sql.Expression, transform.TreeIdentity, error) {
-		s, ok := e.(*plan.Subquery)
-		if !ok {
-			return e, transform.SameTree, nil
+// analyzeSubqueryAlias runs analysis on the specified subquery alias, |sqa|. The |finalize| parameter indicates if this is
+// the final run of the analyzer on the query before execution, which means all rules, starting from the default-rules
+// batch are processed, otherwise only the once-before-default batch of rules is processed for all other non-final passes.
+func analyzeSubqueryAlias(ctx *sql.Context, a *Analyzer, node sql.Node, sqa *plan.SubqueryAlias, scope *Scope, sel RuleSelector, finalize bool) (sql.Node, transform.TreeIdentity, error) {
+	subScope := newScopeWithDepth(scope.RecursionDepth() + 1)
+	if scope != nil && len(scope.nodes) > 0 {
+		// As of MySQL 8.0.14, MySQL provides OUTER scope visibility to derived tables. Unlike LATERAL scope visibility, which
+		// gives a derived table visibility to the adjacent expressions where the subquery is defined, OUTER scope visibility
+		// gives a derived table visibility to the OUTER scope where the subquery is defined.
+		// https://dev.mysql.com/blog-archive/supporting-all-kinds-of-outer-references-in-derived-tables-lateral-or-not/
+		// We don't include the current inner node so that the outer scope nodes are still present, but not the lateral nodes
+		if scope.CurrentNodeIsSubqueryExpression {
+			sqa.OuterScopeVisibility = true
```

Comment:
Probably prefer a method on Scope for this logic (ditto the other place this is done)